dbms/src/Flash/Coprocessor/DAGContext.h

```diff
@@ -328,14 +341,22 @@ class DAGContext
     TableLockHolders table_locks;
     /// profile_streams_map is a map that maps from executor_id to profile BlockInputStreams.
     std::unordered_map<String, BlockInputStreams> profile_streams_map;
+
+    /// executor_id, ExecutorProfileInfo
+    std::unordered_map<String, ExecutorProfile> pipeline_profiles_map;
+
     /// executor_id_to_join_id_map is a map that maps executor id to all the join executor id of itself and all its children.
     std::unordered_map<String, std::vector<String>> executor_id_to_join_id_map;
     /// join_execute_info_map is a map that maps from join_probe_executor_id to JoinExecuteInfo
     /// DAGResponseWriter / JoinStatistics gets JoinExecuteInfo through it.
     std::unordered_map<std::string, JoinExecuteInfo> join_execute_info_map;
-    /// profile_streams_map is a map that maps from executor_id (table_scan / exchange_receiver) to BlockInputStreams.
+    /// inbound_io_input_streams_map is a map that maps from executor_id (table_scan / exchange_receiver) to BlockInputStreams.
     /// BlockInputStreams contains ExchangeReceiverInputStream, CoprocessorBlockInputStream and local_read_input_stream etc.
     std::unordered_map<String, BlockInputStreams> inbound_io_input_streams_map;
+
+    /// io_source_ops_map is a map that maps from executor_id (table_scan / exchange_receiver) to SourceOps.
+    std::unordered_map<String, SourceOps> io_source_ops_map;
```

Comment:
how about just holding the executor_id of `io_input/io_source`, and then we don't need to hold two maps `inbound_io_input_streams_map` and `io_source_ops_map` here.